iroh-resolver/src/unixfs_builder.rs

```diff
@@ -1,55 +1,172 @@
-use std::{fmt::Debug, path::Path, pin::Pin};
+use std::{
+    fmt::Debug,
+    path::{Path, PathBuf},
+    pin::Pin,
+};
 
 use anyhow::{anyhow, ensure, Result};
+use async_recursion::async_recursion;
 use async_trait::async_trait;
-use bytes::Bytes;
+use bytes::{Bytes, BytesMut};
 use cid::{multihash::MultihashDigest, Cid};
-use futures::{Stream, StreamExt};
+use futures::{stream::LocalBoxStream, Stream, StreamExt};
 use iroh_rpc_client::Client;
 use prost::Message;
 use tokio::io::AsyncRead;
 
 use crate::{
+    balanced_tree::TreeBuilder,
     chunker::{self, Chunker, DEFAULT_CHUNK_SIZE_LIMIT},
     codecs::Codec,
     unixfs::{dag_pb, unixfs_pb, DataType, Node, UnixfsNode},
 };
 
+// CID_LEN is the length of a cid
+const CID_LEN: usize = 64;
+
+// the PBNode.Data field may contain some data, this buffer is to ensure that
+// the contents of that data field + the size of the encoded protobuf links
+// will not be larger than the chunk size
+const BUFFER: usize = 1024;
+
+#[derive(Debug, PartialEq)]
+enum DirectoryType {
+    Basic,
+    // TODO: writing hamt sharding not yet implemented
+    Hamt,
+}
+
+/// Entry is the kind of entry in a directory can be either a file or a
+/// folder (if recursive directories are allowed)
+#[derive(Debug)]
+enum Entry {
+    File(File),
+    Directory(Directory),
+}
+
 /// Construct a UnixFS directory.
-#[derive(Debug, Default)]
+#[derive(Debug)]
 pub struct DirectoryBuilder {
+    path: Option<PathBuf>,
     name: Option<String>,
-    files: Vec<File>,
+    entries: Vec<Entry>,
+    typ: DirectoryType,
+    // estimated_size is used to compare the size of the directory
+    // to the chunk size.
+    // When the number of links in a directory get so large that the
+    // links themselves no longer fit into one chunk
+    // we must use HAMT sharding to represent that directory.
+    // max size that the directory can be before it should be created
+    // as a HAMT, rather than a Basic directory
+    estimated_size: usize,
+    chunker: Chunker,
+}
+
+impl Default for DirectoryBuilder {
+    fn default() -> Self {
+        Self {
+            path: None,
+            name: None,
+            entries: Default::default(),
+            typ: DirectoryType::Basic,
+            estimated_size: 0,
+            chunker: Chunker::fixed_size(),
```

Comment:
Okay, so this also feels a bit odd.

the go impl just assumes chunks are 2MB.

Adding a `chunker` here to give some semblance of configurability. But it also feels odd because this chunker is potentially independant of any chunker being used to determine the chunk size of any files belonging in this directory. This might be fine. The reason we need to pay attention to chunk size here is to calculate whether or not we need a HAMT, not to ensure that all of our chunks are uniform. But it seemed odd to me so I thought I'd highlight it.

Reply:
a directory doesn’t use a chunker, lets chat about this